src/meshlink.c

```diff
@@ -1105,68 +825,119 @@ static struct timeval idle(event_loop_t *loop, void *data) {
-		meshlink_errno = MESHLINK_ESTORAGE;
+	if(check_port(mesh) == 0) {
+		meshlink_errno = MESHLINK_ENETWORK;
 		return false;
 	}
 
-	snprintf(filename, sizeof(filename), "%s" SLASH "meshlink.conf", mesh->confbase);
+	/* Create a node for ourself */
 
-	if(!access(filename, F_OK)) {
-		logger(mesh, MESHLINK_DEBUG, "Configuration file %s already exists!\n", filename);
-		meshlink_errno = MESHLINK_EEXIST;
+	mesh->self = new_node();
+	mesh->self->name = xstrdup(mesh->name);
+	mesh->self->devclass = mesh->devclass;
+	mesh->self->ecdsa = ecdsa_set_public_key(ecdsa_get_public_key(mesh->private_key));
+
+	if(!create_initial_config_files(mesh)) {
 		return false;
 	}
 
-	FILE *f = fopen(filename, "w");
+	return true;
+}
 
-	if(!f) {
-		logger(mesh, MESHLINK_DEBUG, "Could not create file %s: %s\n", filename, strerror(errno));
+static bool meshlink_read_config(meshlink_handle_t *mesh) {
+	// Open the configuration file and lock it
+	if(!main_config_lock(mesh)) {
+		logger(NULL, MESHLINK_ERROR, "Cannot lock main config file\n");
 		meshlink_errno = MESHLINK_ESTORAGE;
 		return false;
 	}
 
-	fprintf(f, "Name = %s\n", mesh->name);
-	fclose(f);
+	config_t config;
 
-	if(!ecdsa_keygen(mesh)) {
-		meshlink_errno = MESHLINK_EINTERNAL;
-		unlink(filename);
+	if(!main_config_read(mesh, &config)) {
+		logger(NULL, MESHLINK_ERROR, "Could not read main configuration file!");
 		return false;
 	}
 
-	if(check_port(mesh) == 0) {
-		meshlink_errno = MESHLINK_ENETWORK;
-		unlink(filename);
+	packmsg_input_t in = {config.buf, config.len};
+	const void *private_key;
+	const void *invitation_key;
+
+	uint32_t version = packmsg_get_uint32(&in);
+	char *name = packmsg_get_str_dup(&in);
+	uint32_t private_key_len = packmsg_get_bin_raw(&in, &private_key);
+	uint32_t invitation_key_len = packmsg_get_bin_raw(&in, &invitation_key);
+	uint16_t myport = packmsg_get_uint16(&in);
+
+	if(!packmsg_done(&in) || version != MESHLINK_CONFIG_VERSION || private_key_len != 96 || invitation_key_len != 96) {
+		logger(NULL, MESHLINK_ERROR, "Error parsing main configuration file!");
+		free(name);
+		config_free(&config);
+		return false;
+	}
+
+#if 0
+
+	// TODO: check this?
+	if(mesh->name && strcmp(mesh->name, name)) {
+		logger(NULL, MESHLINK_ERROR, "Configuration is for a different name (%s)!", name);
+		meshlink_errno = MESHLINK_ESTORAGE;
+		free(name);
+		config_free(&config);
+		return false;
+	}
+
+#endif
+
+	free(mesh->name);
+	mesh->name = name;
+	xasprintf(&mesh->myport, "%u", myport);
+	mesh->private_key = ecdsa_set_private_key(private_key);
+	mesh->invitation_key = ecdsa_set_private_key(invitation_key);
+	config_free(&config);
+
+	/* Create a node for ourself and read our host configuration file */
+
+	mesh->self = new_node();
+	mesh->self->name = xstrdup(name);
+	mesh->self->devclass = mesh->devclass;
+
+	if(!node_read_public_key(mesh, mesh->self)) {
+		logger(NULL, MESHLINK_ERROR, "Could not read our host configuration file!");
+		free_node(mesh->self);
+		mesh->self = NULL;
```

Comment:
All of the host node's config files has to be loaded back into respective node handles before returning. 
I found that we only have `load_node()` and the old function `load_all_nodes()` has been removed somehow.

Reply:
This is done in `setup_myself()` in `net_setup.c` via the function `config_scan_all(mesh, load_node)`.